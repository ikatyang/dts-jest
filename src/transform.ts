import * as ts from 'typescript';
import { Group } from './definitions';
import { create_group_expression } from './utils/create-group-expression';
import { create_setup_expression } from './utils/create-setup-expression';
import { create_test_expression } from './utils/create-test-expression';
import { create_triggers } from './utils/create-triggers';

export const transform = (
  source_text: string,
  source_filename: string,
  _jest_config?: any,
) => {
  const source_file = ts.createSourceFile(
    source_filename,
    source_text,
    ts.ScriptTarget.Latest,
    false,
  );

  let last_group: Group | undefined;
  const triggers = create_triggers(source_file);
  return triggers
    .reduce<string[]>((transformed_line_texts, trigger, index) => {
      const is_diff_group = trigger.group !== last_group;

      write_group_close_if_preset(is_diff_group && last_group !== undefined);
      write_group_open_if_preset(is_diff_group);
      write(create_test_expression(trigger), true);
      write_group_close_if_preset(
        trigger.group !== undefined && index === triggers.length - 1,
      );

      last_group = trigger.group;
      return transformed_line_texts;

      function write(str: string, trailing_semicolon: boolean) {
        transformed_line_texts[trigger.line] +=
          str + (trailing_semicolon ? ';' : '');
      }
      function write_group_open_if_preset(value: boolean) {
        if (!value) {
          return;
        }
        write(create_group_expression(trigger.group!, { type: 'open' }), false);
      }
      function write_group_close_if_preset(value: boolean) {
        if (!value) {
          return;
        }
        write(create_group_expression(trigger.group!, { type: 'close' }), true);
      }
    }, source_text.split('\n').map((_, index) => (index !== 0 ? '' : `${create_setup_expression(triggers)};`)))
    .join('\n');
};
