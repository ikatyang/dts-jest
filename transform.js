exports.process = require('./lib/index').transform;
