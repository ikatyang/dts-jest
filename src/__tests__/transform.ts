import * as fs from 'fs';
import * as path from 'path';
import {transform} from '../transform';

const transform_fixture = (relative_path: string) => {
  const source_filename = path.resolve(__dirname, relative_path);
  const source_text = fs.readFileSync(source_filename, 'utf8');
  return transform(source_text, source_filename);
};

it('should transform correctly', () => {
  expect(transform_fixture('../../fixtures/transform/general.ts')).toMatchSnapshot();
});

it('should transform correctly with template token', () => {
  expect(transform_fixture('../../fixtures/transform/template-token.ts')).toMatchSnapshot();
});

it('should transform correctly with indented expression', () => {
  expect(transform_fixture('../../fixtures/transform/indented-expression.ts')).toMatchSnapshot();
});

it('should throw error with unexpected flag', () => {
  expect(() => transform_fixture('../../fixtures/transform/unexpected.ts')).toThrowError();
});

it('should throw error with unattachable trigger(s)', () => {
  expect(() => transform_fixture('../../fixtures/transform/unattachable.ts')).toThrowErrorMatchingSnapshot();
});
