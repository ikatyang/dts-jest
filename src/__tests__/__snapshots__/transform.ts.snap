// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should throw error with unattachable trigger(s) 1`] = `
"Unattachable trigger(s) detected:

  fixtures/transform/unattachable.ts:1
  fixtures/transform/unattachable.ts:3
  fixtures/transform/unattachable.ts:5
  fixtures/transform/unattachable.ts:7"
`;

exports[`should transform correctly 1`] = `
"var _dts_jest_runtime_;beforeAll(function (done) {require(\\"dts-jest\\").setup(module, (function () {try {return _dts_jest_;} catch (e) {return {};}})(), {\\"4\\":\\"T.none()\\",\\"7\\":\\"T.none()\\",\\"10\\":\\"T.show()\\",\\"13\\":\\"T.show()\\",\\"16\\":\\"T.skip()\\",\\"19\\":\\"T.skip()\\",\\"22\\":\\"T.only()\\",\\"25\\":\\"T.only()\\",\\"29\\":\\"T.in_block()\\"}, function () {_dts_jest_runtime_ = arguments[0];done();});});



test(\\"T.none()\\", function () {expect(_dts_jest_runtime_.snapshot(4)).toMatchSnapshot();});


test(\\"description-none\\", function () {expect(_dts_jest_runtime_.snapshot(7)).toMatchSnapshot();});


test(\\"T.show()\\", function () {console.log(_dts_jest_runtime_.report(10));});


test(\\"description-show\\", function () {console.log(_dts_jest_runtime_.report(13));});


test.skip(\\"T.skip()\\", function () {expect(_dts_jest_runtime_.snapshot(16)).toMatchSnapshot();});


test.skip(\\"description-skip\\", function () {expect(_dts_jest_runtime_.snapshot(19)).toMatchSnapshot();});


test.only(\\"T.only()\\", function () {expect(_dts_jest_runtime_.snapshot(22)).toMatchSnapshot();});


test.only(\\"description-only\\", function () {expect(_dts_jest_runtime_.snapshot(25)).toMatchSnapshot();});



test(\\"T.in_block()\\", function () {expect(_dts_jest_runtime_.snapshot(29)).toMatchSnapshot();});


"
`;

exports[`should transform correctly with template token 1`] = `
"var _dts_jest_runtime_;beforeAll(function (done) {require(\\"dts-jest\\").setup(module, (function () {try {return _dts_jest_;} catch (e) {return {};}})(), {\\"4\\":\\"R.keysIn(f)\\",\\"10\\":\\"R.keysIn(f)\\",\\"26\\":\\"R.toString(new Point(1, 2))\\"}, function () {_dts_jest_runtime_ = arguments[0];done();});});



test(\\"R.keysIn(f)\\", function () {expect(_dts_jest_runtime_.snapshot(4)).toMatchSnapshot();});





test(\\"R.keysIn(f)\\", function () {expect(_dts_jest_runtime_.snapshot(10)).toMatchSnapshot();});















test(\\"R.toString(new Point(1, 2))\\", function () {expect(_dts_jest_runtime_.snapshot(26)).toMatchSnapshot();});

"
`;
